### PR TITLE
Support gdbstub kernal debugging

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
Support gdbstub kernal debugging, need merge with https://github.com/apache/nuttx-apps/pull/2224 together.
## Impact
gdbstub
## Testing
use demo for test

A gdbstub is built into nuttx, which implements an almost complete GDB rsp protocol. You can use the serial port to debug the device, single-step, breakpoint, watchpoint and other debugging function.

To support other architectures, you need to support the` up_debugpoint_add/up_debugpoint_remove` API.

### CODE REVIEW

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 1]
```diff
@@ -0,0 +1,673 @@
...
+        {
+          uint32_t fun_type = (fun & DWT_FUNCTION_FUNCTION_MASK) >>
+                              DWT_FUNCTION_FUNCTION_SHIFT;
+
+          addr = getreg32(DWT_COMP0 + 16 * i);
+          switch (fun_type)
+            {
+              case DWT_FUNCTION_WATCHPOINT_RO:
+                  type = DEBUGPOINT_WATCHPOINT_RO;
+                  break;
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:271]:
```suggestion
              case DWT_FUNCTION_WATCHPOINT_RO:
                type = DEBUGPOINT_WATCHPOINT_RO;
                break;
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 2]
```diff
@@ -0,0 +1,673 @@
...
+              case DWT_FUNCTION_WATCHPOINT_WO:
+                  type = DEBUGPOINT_WATCHPOINT_WO;
+                  break;
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:274]:
```suggestion
              case DWT_FUNCTION_WATCHPOINT_WO:
                type = DEBUGPOINT_WATCHPOINT_WO;
                break;
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 3]
```diff
@@ -0,0 +1,673 @@
...
+              case DWT_FUNCTION_WATCHPOINT_RW:
+                  type = DEBUGPOINT_WATCHPOINT_RW;
+                  break;
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:277]:
```suggestion
              case DWT_FUNCTION_WATCHPOINT_RW:
                type = DEBUGPOINT_WATCHPOINT_RW;
                break;
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 4]
```diff
@@ -0,0 +1,673 @@
...
+{
+  uint32_t num = ARM_DWT_NUM();
+  uint32_t addr = 0;
+  uint32_t i;
+  int type = 0;
+
+  for (i = 0; i < num; i++)
+    {
+      uint32_t fun = getreg32(DWT_FUNCTION0 + 16 * i);
+      if (fun & DWT_FUNCTION_MATCHED_MASK)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:261]: Optional
```suggestion
      if ((fun & DWT_FUNCTION_MATCHED_MASK) != 0)
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 5]
```diff
@@ -0,0 +1,673 @@
...
+
+static void arm_breakpoint_match(uint32_t pc)
+{
+  uint32_t i;
+
+  for (i = 0; i < ARM_DEBUG_MAX; i++)
+    {
+      if (g_arm_debug[i].type == DEBUGPOINT_BREAKPOINT &&
+          g_arm_debug[i].addr == (void *)pc &&
+          g_arm_debug[i].callback)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:423]: Optional
```suggestion
          g_arm_debug[i].callback != NULL)
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 6]
```diff
@@ -0,0 +1,673 @@
...
+ ****************************************************************************/
+
+static void arm_steppoint_match(void)
+{
+  uint32_t i;
+
+  for (i = 0; i < ARM_DEBUG_MAX; i++)
+    {
+      if (g_arm_debug[i].type == DEBUGPOINT_STEPPOINT &&
+          g_arm_debug[i].callback)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:477]: Optional
```suggestion
          g_arm_debug[i].callback != NULL)
```

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 7]
```diff
@@ -0,0 +1,673 @@
...
+ *   debug events, and resume execution.
+ *
+ ****************************************************************************/
+
+int arm_dbgmonitor(int irq, void *context, void *arg)
+{
+  uint32_t dfsr = getreg32(NVIC_DFAULTS);
+  uint32_t *regs = (uint32_t *)context;
+
+  if (dfsr & NVIC_DFAULTS_HALTED)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:657]: Optional: add `!= 0`

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 8]
```diff
@@ -0,0 +1,673 @@
...
+    {
+      arm_steppoint_match();
+    }
+
+  if (dfsr & NVIC_DFAULTS_BKPT)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:662]: same

File: arch/arm/src/armv7-m/arm_dbgmonitor.c [hunk 9]
```diff
@@ -0,0 +1,673 @@
...
+    {
+      arm_breakpoint_match(regs[REG_PC]);
+    }
+
+  if (dfsr & NVIC_DFAULTS_DWTTRAP)
```
Comment (pkarashchenko) [inline arch/arm/src/armv7-m/arm_dbgmonitor.c:667]: same

File: arch/arm/src/armv8-m/arm_dbgmonitor.c [hunk 10]
```diff
@@ -0,0 +1,700 @@
+/****************************************************************************
```
Comment (pkarashchenko) [inline arch/arm/src/armv8-m/arm_dbgmonitor.c:1]: All the same comments as for armv7

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 11]
```diff
@@ -1366,13 +1365,260 @@ static int gdb_thread_context(FAR struct gdb_state_s *state)
...
+    }
+
+  switch (type)
+    {
+      case 0: /* just use hardware break */
+      case 1:
+        type = DEBUGPOINT_BREAKPOINT;
+        break;
+      case 2:
+          type = DEBUGPOINT_WATCHPOINT_WO;
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1536]:
```suggestion
        type = DEBUGPOINT_WATCHPOINT_WO;
```

File: libs/libc/gdbstub/lib_gdbstub.c [hunk 12]
```diff
@@ -1366,13 +1365,260 @@ static int gdb_thread_context(FAR struct gdb_state_s *state)
...
+        break;
+      case 3:
+          type = DEBUGPOINT_WATCHPOINT_RO;
```
Comment (pkarashchenko) [inline libs/libc/gdbstub/lib_gdbstub.c:1539]:
```suggestion
        type = DEBUGPOINT_WATCHPOINT_RO;
```

**Comment by xiaoxiang781216** [comment]
@anjiahao1 please fix the warning:
```
gdbstub/lib_gdbstub.c: In function 'gdb_process':
Error: gdbstub/lib_gdbstub.c:1837:1: error: label 'out' defined but not used [-Werror=unused-label]
 1837 | out:
      | ^~~
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by raiden00pl** [comment]
hi @anjiahao1 it looks like https://github.com/apache/nuttx/pull/11350/commits/a79b98b0516868f3473e5955b78f29c16f5eec83 has broken gdb breakpoints for nrf52 when used with JLink. I didn't look further for the problem, but I see that the breakpoints don't work now.

**Comment by anjiahao1** [comment]
> hi @anjiahao1 it looks like [a79b98b](https://github.com/apache/nuttx/commit/a79b98b0516868f3473e5955b78f29c16f5eec83) has broken gdb breakpoints for nrf52 when used with JLink. I didn't look further for the problem, but I see that the breakpoints don't work now.
> 
> EDIT: it looks like one of the options I'm using is incompatible with this PR. Breakpoints work with a basic NSH config
> 
> EDIT2: this PR and `CONFIG_DEBUG_FEATURES=y` breaks breakpoints at least for nrf52

According to the ARM manual, when using swo/jtag for debugging, the internal DWT and FPB modules are disabled.

**Comment by raiden00pl** [comment]
@anjiahao1 
`b-g431b-esc1` is also broken. Tested with JLink and STLink, SDW only. So it's possible that breakpoints for more upstream boards are broken now.

**Comment by raiden00pl** [comment]
@anjiahao1 
https://github.com/apache/nuttx/blob/9483729328d4aa2986a94aa38668495fbe113b93/arch/arm/src/armv7-m/arm_dbgmonitor.c#L95

this line resets all breakpoints set before we running our program in debugger which is wrong.
A normal debugging session looks like this: 1. reset target, 2. set some breakpoints, 3. run target. If our breakpoints are set after `arm_fpb_init` call then it will never work.

 If this behavior is required for gdbstub then this must be controlled by Kconfig option which is *off* by default.

**Comment by raiden00pl** [comment]
> According to the ARM manual, when using swo/jtag for debugging, the internal DWT and FPB modules are disabled.
 
 Could you provide the source in manual ? Maybe this is a vendor-dependent feature? What about SWD ? I don't have a JTAG setup nearby to check right now to check if it works.

**Comment by anjiahao1** [comment]
> > According to the ARM manual, when using swo/jtag for debugging, the internal DWT and FPB modules are disabled.
> 
> Could you provide the source in manual ? Maybe this is a vendor-dependent feature? What about SWD ? I don't have a JTAG setup nearby to check right now if it works.

this [pr](https://github.com/apache/nuttx/pull/11413) fix it. I misunderstood before. If `jtag` or `swo` is connected, the `debug monito`r interrupt is disabled, not `fpb` and `dwt`. Now the new pr fixes this problem and can handle both situations.

**Pull request merged**